src/style-spec/expression/definitions/within.js

```diff
@@ -0,0 +1,273 @@
+    if (bbox1[0] <= bbox2[0]) return false;
+    if (bbox1[2] >= bbox2[2]) return false;
+    if (bbox1[1] <= bbox2[1]) return false;
+    if (bbox1[3] >= bbox2[3]) return false;
+    return true;
+}
+
+function getLngLatPoint(coord: Point, canonical) {
+    const tilesAtZoom = Math.pow(2, canonical.z);
+    const x = (coord.x / EXTENT + canonical.x) / tilesAtZoom;
+    const y = (coord.y / EXTENT + canonical.y) / tilesAtZoom;
+    const p = new MercatorCoordinate(x, y).toLngLat();
+
+    return [p.lng, p.lat];
+}
+
+function getLngLatPoints(line, canonical) {
+    const coords = [];
+    for (let i = 0; i < line.length; ++i) {
+        coords.push(getLngLatPoint(line[i], canonical));
+    }
+    return coords;
+}
+
+function onBoundary(p, p1, p2) {
+    const x1 = p[0] - p1[0];
+    const y1 = p[1] - p1[1];
+    const x2 = p[0] - p2[0];
+    const y2 = p[1] - p2[1];
+    return (x1 * y2 - x2 * y1 === 0) && (x1 * x2 <= 0) && (y1 * y2 <= 0);
+}
+
+function rayIntersect(p, p1, p2) {
+    return ((p1[1] > p[1]) !== (p2[1] > p[1])) && (p[0] < (p2[0] - p1[0]) * (p[1] - p1[1]) / (p2[1] - p1[1]) + p1[0]);
+}
+
+// ray casting algorithm for detecting if point is in polygon
+function pointWithinPolygon(point, rings) {
+    let inside = false;
+    for (let i = 0, len = rings.length; i < len; i++) {
+        const ring = rings[i];
+        for (let j = 0, len2 = ring.length; j < len2 - 1; j++) {
+            if (onBoundary(point, ring[j], ring[j + 1])) return false;
+            if (rayIntersect(point, ring[j], ring[j + 1])) inside = !inside;
+        }
+    }
+    return inside;
+}
+
+function pointWithinPolygons(point, polygons) {
+    if (polygons.type === 'Polygon') {
+        return pointWithinPolygon(point, polygons.coordinates);
+    }
+    for (let i = 0; i < polygons.coordinates.length; i++) {
+        if (!pointWithinPolygon(point, polygons.coordinates[i])) return false;
+    }
+    return true;
+}
+
+// a, b are end points for line segment1, c and d are end points for line segment2
+function lineIntersectLine(a, b, c, d) {
+    const perp = (v1, v2) => { return (v1[0] * v2[1] - v1[1] * v2[0]); };
+
+    // check if two segments are parallel or not
+    // precondition is end point a, b is inside polygon, if line a->b is
+    // parallel to polygon edge c->d, then a->b won't intersect with c->d
+    const vectorP = [b[0] - a[0], b[1] - a[1]];
+    const vectorQ = [d[0] - c[0], d[1] - c[1]];
+    if (perp(vectorQ, vectorP) === 0) return false;
+
+    // check if p1 and p2 are in different sides of line segment q1->q2
+    const twoSided = (p1, p2, q1, q2)  => {
+        // q1->p1 (x1, y1), q1->p2 (x2, y2), q1->q2 (x3, y3)
+        const x1 = p1[0] - q1[0];
+        const y1 = p1[1] - q1[1];
+        const x2 = p2[0] - q1[0];
+        const y2 = p2[1] - q1[1];
+        const x3 = q2[0] - q1[0];
+        const y3 = q2[1] - q1[1];
+        if ((x1 * y3 - x3 * y1) * (x2 * y3 - x3 * y2) < 0) return true;
+        return false;
+    };
```

Comment:
All the anonymous functions here (`perp`, `twoSided`) should be cut out outside because JS will allocate the functions on every `lineIntersectLine` call, unlike C++ that usually optimizes lambdas away.

Reply:
Thanks 🙇‍♀